src/engine/cuecontrol.cpp

```diff
@@ -754,12 +762,56 @@ void CueControl::cueDenon(double v) {
     }
 }
 
+void CueControl::cuePlay(double v) {
+    // This is how CUP button works:
+    // If playing, press to go to cue and stop.
+	// If stopped, press to set as cue point.
+	// On release, start playing from cue point.
+
+
+    QMutexLocker lock(&m_mutex);
+    bool playing = (m_pPlayButton->get() == 1.0);
+
+    // pressed
+    if(v){
+		if (playing) {
+			m_bPreviewing = false;
+			m_pPlayButton->set(0.0);
+
+			// Need to unlock before emitting any signals to prevent deadlock.
+			lock.unlock();
+
+			seekAbs(m_pCuePoint->get());
+		} else if (!isTrackAtCue() && getCurrentSample() <= getTotalSamples()) {
+			// Pause not at cue point and not at end position
+			cueSet(v);
+			// Just in case.
+			m_bPreviewing = false;
+
+			// If quantize is enabled, jump to the cue point since it's not
+			// necessarily where we currently are
+			if (m_pQuantizeEnabled->get() > 0.0) {
+				lock.unlock();  // prevent deadlock.
+				// Enginebuffer will quantize more exactly than we can.
+				seekAbs(m_pCuePoint->get());
+			}
+		}
+    } else if (isTrackAtCue()){
+        m_bPreviewing = false;
+        m_pPlayButton->set(1.0);
+        lock.unlock();
+
+    }
+}
+
 void CueControl::cueDefault(double v) {
     double cueMode = m_pCueMode->get();
     // Decide which cue implementation to call based on the user preference
     if (cueMode == CUE_MODE_DENON || cueMode == CUE_MODE_NUMARK) {
         cueDenon(v);
-    } else {
+    }else if(cueMode == CUE_MODE_CUP){
+        cuePlay(v);
```

Comment:
Will you use this mode?  I am just asking, because I have found more occurrences where CUP replaces Play. 
http://forum.djtechtools.com/showthread.php?t=17634&p=156018&viewfull=1#post156018

If it replaces the CUE button, than the cue set feature from above makes sense, otherwise not. 
mmmm that weird, do we actually need two CUP buttons, On with cue set function and on without?